L1Trigger/L1TTrackMatch/plugins/DisplacedVertexProducer.cc

```diff
@@ -0,0 +1,221 @@
+    x_vtx = ix2_x;
+    y_vtx = ix2_y;
+    z_vtx = (ix2_z1+ix2_z2)/2;
+    return 2;
+  }
+  else{
+    if(ix1_delz<ix2_delz){
+      x_vtx = ix1_x;
+      y_vtx = ix1_y;
+      z_vtx = (ix1_z1+ix1_z2)/2;
+      return 3;
+    }
+    else{
+      x_vtx = ix2_x;
+      y_vtx = ix2_y;
+      z_vtx = (ix2_z1+ix2_z2)/2;
+      return 3;
+    }
+  }
+  return 4;
+}
+
+DisplacedVertexProducer::DisplacedVertexProducer(const edm::ParameterSet& iConfig)
+  : ttTrackMCTruthToken_(consumes<TTTrackAssociationMap<Ref_Phase2TrackerDigi_> >(iConfig.getParameter<edm::InputTag>("mcTruthTrackInputTag"))),
+    trackToken_(consumes<std::vector<TTTrack<Ref_Phase2TrackerDigi_> > >(iConfig.getParameter<edm::InputTag>("l1TracksInputTag"))),
+    outputTrackCollectionName_(iConfig.getParameter<std::string>("l1TrackVertexCollectionName")),
+    ONNXmodel_(iConfig.getParameter<std::string>("ONNXmodel")),
+    ONNXInputName_(iConfig.getParameter<std::string>("ONNXInputName")),
+    featureNames_(iConfig.getParameter<std::vector<std::string>>("featureNames"))
+{
+  //--- Define EDM output to be written to file (if required)
+  produces<l1t::DisplacedTrackVertexCollection>(outputTrackCollectionName_);
+  runTime_ = std::make_unique<cms::Ort::ONNXRuntime>(this->ONNXmodel_);
+}
+
+void DisplacedVertexProducer::produce(edm::StreamID, edm::Event& iEvent, const edm::EventSetup& iSetup) const {
+  edm::Handle<TTTrackAssociationMap<Ref_Phase2TrackerDigi_> > MCTruthTTTrackHandle;
+  iEvent.getByToken(ttTrackMCTruthToken_, MCTruthTTTrackHandle);
+  edm::Handle<std::vector<TTTrack<Ref_Phase2TrackerDigi_> > > TTTrackHandle;
+  iEvent.getByToken(trackToken_, TTTrackHandle);
+  std::vector<TTTrack<Ref_Phase2TrackerDigi_> >::const_iterator iterL1Track;
+  int this_l1track = 0;
+  std::vector<Track_Parameters> selectedTracks;
+  std::vector<edm::Ptr<TrackingParticle>> selectedTPs;
+  
+  for (iterL1Track = TTTrackHandle->begin(); iterL1Track != TTTrackHandle->end(); iterL1Track++) {
+    edm::Ptr<TTTrack<Ref_Phase2TrackerDigi_> > l1track_ptr(TTTrackHandle, this_l1track);
+    this_l1track++;
+    float pt = iterL1Track->momentum().perp();
+    float eta = iterL1Track->momentum().eta();
+    float phi = iterL1Track->momentum().phi();
+    float z0 = iterL1Track->z0();  //cm
+    float x0 = iterL1Track->POCA().x();
+    float y0 = iterL1Track->POCA().y();
+    float d0 = -x0 * sin(phi) + y0 * cos(phi);
+    float rinv = iterL1Track->rInv();
+    float chi2 = iterL1Track->chi2Red();
+    float chi2rphi = iterL1Track->chi2XYRed();
+    float chi2rz = iterL1Track->chi2ZRed();
+    float bendchi2 = iterL1Track->stubPtConsistency();
+    float MVA1 = iterL1Track->trkMVA1();
+    float MVA2 = iterL1Track->trkMVA2();
+    std::vector<edm::Ref<edmNew::DetSetVector<TTStub<Ref_Phase2TrackerDigi_> >, TTStub<Ref_Phase2TrackerDigi_> > > stubRefs = iterL1Track->getStubRefs();
+    int nstub = (int)stubRefs.size();
+    if( chi2rz<3.0 && MVA2>0.2 && MVA1>0.2 && pt>3.0 && fabs(eta)<2.4){
```

Comment:
I would recommend moving these track selection criteria to the python config to avoid the hard coding of these criteria here or using the track selection producer

Reply:
I second using the track selection producer as this is much cleaner. You will just have to create another config in the [python config](https://github.com/cms-l1t-offline/cmssw/blob/phase2-l1t-integration-14_0_0_pre3/L1Trigger/L1TTrackMatch/python/l1tTrackSelectionProducer_cfi.py) and add whatever varibles are not already there (like MVA2), and then pass this track collection into the displaced vertex config [here](https://github.com/ryanm124/cmssw/blob/displacedVertexBDTSim/L1Trigger/L1TTrackMatch/python/DisplacedVertexProducer_cfi.py#L4).

Reply:
I'm looking at the track selector and I'm wondering if using this is in fact cleaner than simply moving the parameters to a python config as Christopher suggested. The selector producer seems to be made for simpler, single variable cuts while I have several selections that depend on two variables such as the MVA score of displaced (d0>1cm) tracks. The selector also is cutting on all defined variables regardless of which ones the user has overwritten so there are many wasted operations such as checking if reducedBendChi2MaxNstub5<999.9 which would always be true.

Reply:
I would agree with @Chriisbrown and @cgsavard. You should add these to the python configuration of this module so the values of the cuts can be set there, but keep the logic here. I don't see where wasted checks and operations would happen there?

Reply:
@aloeliger For example, [reducedChi2RZMaxNstub4](https://github.com/cms-l1t-offline/cmssw/blob/9766868c6f413c0f1dcc83f803a3186393cd0678/L1Trigger/L1TTrackMatch/python/l1tTrackSelectionProducer_cfi.py#L17) is defined which I'm currently not cutting on. This parameter is used as part of the[ chi2nstub](https://github.com/cms-l1t-offline/cmssw/blob/9766868c6f413c0f1dcc83f803a3186393cd0678/L1Trigger/L1TTrackMatch/plugins/L1TrackSelectionProducer.cc#L548) selection which is used for every track [here](https://github.com/cms-l1t-offline/cmssw/blob/9766868c6f413c0f1dcc83f803a3186393cd0678/L1Trigger/L1TTrackMatch/plugins/L1TrackSelectionProducer.cc#L560). So this track selection producer seems to cut on every defined variable even though for many of the default values such as this one the cut is essentially a Boolean true

Reply:
Right, but couldn't you configure the cuts required from [L1Trigger/L1TTrackMatch/python/DisplacedVertexProducer_cfi.py](https://github.com/cms-l1t-offline/cmssw/pull/1257/files#diff-de035f9dfa2af663c1bd8d1d131605afd072d09a96d1fd6076e22a469610cd6b)?

Reply:
My understanding from @cgsavard 's comment above was that the track selection is configured using its own [config file](https://github.com/cms-l1t-offline/cmssw/blob/phase2-l1t-integration-14_0_0_pre3/L1Trigger/L1TTrackMatch/python/l1tTrackSelectionProducer_cfi.py), it's then placed in the schedule before my displaced vertexing, and outputs the track collection which is then input to my producer. I could definitely move my cuts to my [DisplacedVertexProducer_cfi.py](https://github.com/cms-l1t-offline/cmssw/pull/1257/files#diff-de035f9dfa2af663c1bd8d1d131605afd072d09a96d1fd6076e22a469610cd6b) but I don't see how that removes the unneeded cuts like [reducedChi2RZMaxNstub4](https://github.com/cms-l1t-offline/cmssw/blob/9766868c6f413c0f1dcc83f803a3186393cd0678/L1Trigger/L1TTrackMatch/python/l1tTrackSelectionProducer_cfi.py#L17) if I use the track selector producer.

Reply:
I see what you are saying about the cuts you don't use, but theoretically on hardware we won't be implementing these cuts and we only set them up like this in the software so that we can use the same track selection module on each physics object. I think it's best to follow the structure all other GTT physics objects are using, therefore using the track selection module.

Reply:
How were these [parameter magnitude sizes](https://github.com/cms-l1t-offline/cmssw/blob/9766868c6f413c0f1dcc83f803a3186393cd0678/L1Trigger/L1TTrackMatch/plugins/L1TrackSelectionProducer.cc#L80) determined? I'm trying to add d0 to this list

Reply:
You shouldn't need to set a magnitude size, this is for when a variable needs to be converted from a track word variable. Since d0 is a variable directly included in the track word, no conversion is needed. You should follow what is being done for [getting the z0](https://github.com/cms-l1t-offline/cmssw/blob/9766868c6f413c0f1dcc83f803a3186393cd0678/L1Trigger/L1TTrackMatch/plugins/L1TrackSelectionProducer.cc#L144-L164) using the proper [d0 functions from the track word class](https://github.com/cms-l1t-offline/cmssw/blob/9766868c6f413c0f1dcc83f803a3186393cd0678/DataFormats/L1TrackTrigger/interface/TTTrack_TrackWord.h#L208).

Reply:
I've modified the displaced vertex producer to use the track selection producer here [63cdfcf](https://github.com/cms-l1t-offline/cmssw/pull/1257/commits/63cdfcf8c0dbba752b18dac750196655f6cf1b35)

Reply:
Does the MC track association have to be redone after using the track selection producer? The truth association seems to not be working now after this change

Reply:
No, it shouldn't change anything except for where track selection is being done. The only thing I can think of is some of the files don't have the truth association included so maybe you are looking at one of those files?

Reply:
I've been using the same file for testing these PR changes and I was able to get the tracking particles before. Now it's giving a null pointer for each track. I looked at the other producers that use the track selection and it seems like they don't use truth information. Has anyone done this before?